go/internal/logutil/logutil_test.go

```diff
@@ -1,61 +1,174 @@
+		t.Skip("unittest not consistent on windows, skipping.")
 	}
+
+	closer, err := u.CaptureStdoutAndStderr()
+
+	logger, cleanup, err := logutil.NewLogger(
+		logutil.NewStdStream("*", "light-console", "stdout"),
+	)
+	require.NoError(t, err)
 	defer cleanup()
 
-	logger.Debug("top debug")
-	logger.Info("top info")
-	logger.Warn("top warn")
-	logger.Error("top error")
-
-	logger.Named("foo").Debug("foo debug")
-	logger.Named("foo").Info("foo info")
-	logger.Named("foo").Warn("foo warn")
-	logger.Named("foo").Error("foo error")
-
-	// Output:
-	// DEBUG	bty               	logutil/logutil_test.go:14	top debug
-	// INFO 	bty               	logutil/logutil_test.go:15	top info
-	// WARN 	bty               	logutil/logutil_test.go:16	top warn
-	// ERROR	bty               	logutil/logutil_test.go:17	top error
-	// DEBUG	bty.foo           	logutil/logutil_test.go:19	foo debug
-	// INFO 	bty.foo           	logutil/logutil_test.go:20	foo info
-	// WARN 	bty.foo           	logutil/logutil_test.go:21	foo warn
-	// ERROR	bty.foo           	logutil/logutil_test.go:22	foo error
+	logger.Info("hello world!")
+	logger.Warn("hello world!")
+	logger.Sync()
+	lines := strings.Split(strings.TrimSpace(closer()), "\n")
+	require.Equal(t, 2, len(lines))
+	require.Equal(t, "INFO \tbty               \tlogutil/logutil_test.go:34\thello world!", lines[0])
+	require.Equal(t, "WARN \tbty               \tlogutil/logutil_test.go:35\thello world!", lines[1])
 }
 
-func Example_logerrors() {
-	logger, cleanup, err := logutil.NewLogger("error:*,-*.bar warn:*.bar", "light-console", "stdout")
-	if err != nil {
-		panic(err)
+func TestTypeRing(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("unittest not consistent on windows, skipping.")
 	}
+
+	closer, err := u.CaptureStdoutAndStderr()
+
+	ring := zapring.New(10 * 1024 * 1024) // 10MB ring-buffer
+	defer ring.Close()
+
+	logger, cleanup, err := logutil.NewLogger(
+		logutil.NewRingStream("*", "light-console", ring),
+	)
 	defer cleanup()
+	require.NoError(t, err)
+
+	logger.Info("hello world!")
+	logger.Warn("hello world!")
+	logger.Sync()
+
+	require.Empty(t, closer())
+
+	r, w := io.Pipe()
+	go func() {
+		_, err := ring.WriteTo(w)
+		require.True(t, err == nil || err == io.EOF)
+		w.Close()
+	}()
+	scanner := bufio.NewScanner(r)
+	scanner.Scan()
+	require.Equal(t, "INFO \tbty               \tlogutil/logutil_test.go:59\thello world!", scanner.Text())
+	scanner.Scan()
+	require.Equal(t, "WARN \tbty               \tlogutil/logutil_test.go:60\thello world!", scanner.Text())
+}
+
+func TestTypeLumberjack(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("unittest not consistent on windows, skipping.")
+	}
+
+	tempdir, err := ioutil.TempDir("", "logutil-lumberjack")
+	require.NoError(t, err)
+
+	closer, err := u.CaptureStdoutAndStderr()
+	logger, cleanup, err := logutil.NewLogger(
+		logutil.NewLumberjackStream("*", "light-console", &lumberjack.Logger{
+			Filename: path.Join(tempdir, "test.log"),
```

Comment:
lumberjack has a built-in logrotate feature, but since it takes a filepath and not a dirpath, we can configure it to be named with `${session-start-time}.log`